### PR TITLE
[api-merge] Generate all api's in single pass.

### PR DESCRIPTION
**Pull request by jpobst**
When we run `make jenkins`, we run `api-merge.exe` N times, where N is the number of Android Platform Levels (currently 12).  Each successive run has to merge all previous levels and then the new level.  

That is, our runs look like this:
- 10 + 15 + 16 + 17 +18 + 19 -> 19
- 10 + 15 + 16 + 17 +18 + 19 + 20 -> 20
- 10 + 15 + 16 + 17 +18 + 19 + 20 + 21 -> 21
- etc.

Instead, we can do a single pass where we do all the merging once, outputting the `api.xml` files in between the merges as needed.

- Merge 10 + 15 + 16 + 17 + 18 + 19
- Write 19
- Merge 20
- Write 20
- etc.

This shaves about 4 minutes off of `make jenkins`:

```
Before: 25:35
After: 21:44
```

Note:
This may not be the best place to insert this process. Should we instead do it somewhere earlier in the `make jenkins` chain, and then the existing targets will skip running because of inputs/outputs?  If so, where should we add it?

### CODE REVIEW

File: build-tools/api-merge/api-merge.cs [hunk 1]
```diff
@@ -77,6 +92,55 @@ public static int Main (string[] args)
...
 
+		static int RunConfigurationFile (string config, string inputDir, string outputDir)
+		{
+			if (!File.Exists (config)) {
+				Console.WriteLine ($"error: config file {config} not found");
+				return 3;
+			}
+
+			var doc = XDocument.Load (config);
+			var inputs = doc.Root.Element ("Inputs").Elements ("File").Select (elem => Tuple.Create (Path.Combine (inputDir, elem.Attribute ("Path").Value), elem.Attribute ("Level").Value)).ToList ();
```
Comment (jonpryor) [inline build-tools/api-merge/api-merge.cs]: Could we please *avoid* `Tuple`?  `ValueTuple<...>` would be fine, so long as the members have name.  Reading `.Item1` makes my eyes bleed.  Or just add a new `struct`!

Comment (jpobst) [reply build-tools/api-merge/api-merge.cs]: Nice, didn't realize named tuples were that easy!

File: src/Mono.Android/Mono.Android.targets [hunk 2]
```diff
@@ -65,13 +65,12 @@
     </ItemGroup>
     <PropertyGroup>
       <ApiMerge>..\..\bin\Build$(Configuration)\api-merge.exe</ApiMerge>
-      <_Profiles>@(_AndroidProfile->'%(Identity)', ' ')</_Profiles>
-      <_Glob>-s '..\..\bin\Build$(Configuration)\api\api-*.xml.in'</_Glob>
-      <_LastProfile>--last-description=..\..\bin\Build$(Configuration)\api\api-$(AndroidPlatformId).xml.in</_LastProfile>
-      <_Out>-o "$(IntermediateOutputPath)mcw\api.xml"</_Out>
+      <_ConfigurationFile>..\..\bin\Build$(Configuration)\merge-configuration.xml</_ConfigurationFile>
+      <_ConfigurationInputBaseDirectory>..\..\bin\Build$(Configuration)\api\</_ConfigurationInputBaseDirectory>
+      <_ConfigurationOutputBaseDirectory>obj\$(Configuration)\</_ConfigurationOutputBaseDirectory>
```
Comment (radekdoulik) [inline src/Mono.Android/Mono.Android.targets]: I think this should be `<_ConfigurationOutputBaseDirectory>$(IntermediateOutputPath)</_ConfigurationOutputBaseDirectory>`

The build now fails with:
```
_GenerateApiDescription:
  Creating directory "obj/Release/monoandroid10/android-29/mcw".
  mono --debug=casts ../../bin/BuildRelease/api-merge.exe -config=../../bin/BuildRelease/merge-configuration.xml -config-input-dir=../../bin/BuildRelease/api/ -config-output-dir=obj/Release/
  api-merge: writing output obj/Release/android-19\mcw\api.xml...
  api-merge: writing output obj/Release/android-20\mcw\api.xml...
  api-merge: writing output obj/Release/android-21\mcw\api.xml...
  api-merge: writing output obj/Release/android-22\mcw\api.xml...
  api-merge: writing output obj/Release/android-23\mcw\api.xml...
  api-merge: writing output obj/Release/android-24\mcw\api.xml...
  api-merge: writing output obj/Release/android-25\mcw\api.xml...
  api-merge: writing output obj/Release/android-26\mcw\api.xml...
  api-merge: writing output obj/Release/android-27\mcw\api.xml...
  api-merge: writing output obj/Release/android-28\mcw\api.xml...
  api-merge: writing output obj/Release/android-29\mcw\api.xml...
  api-merge: writing output obj/Release/android-R\mcw\api.xml...
_GenerateBinding:
  mono --debug=casts "/Users/builder/azdo/_work/3/s/xamarin-android/bin/Release/lib/xamarin.android/xbuild/Xamarin/Android/generator.exe" --public --product-version=7 --api-level=29 -o "obj/Release/monoandroid10/android-29/mcw" --codegen-target=XAJavaInterop1 --fixup=metadata --preserve-enums --enumflags=enumflags --enumfields=map.csv --enummethods=methodmap.csv --enummetadata=obj/Release/monoandroid10/android-29/mcw/enummetadata --apiversions="/Users/builder/Library/Android/sdk/platforms/android-29/data/api-versions.xml" --annotations="/Users/builder/Library/Android/sdk/platform-tools/api/annotations.zip" --assembly="Mono.Android, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null" --type-map-report=obj/Release/monoandroid10/android-29/mcw/type-mapping.txt  --enumdir=obj/Release/monoandroid10/android-29/mcw obj/Release/monoandroid10/android-29/mcw/api.xml
  
  Unhandled Exception:
  System.IO.FileNotFoundException: Could not find file "/Users/builder/azdo/_work/3/s/xamarin-android/src/Mono.Android/obj/Release/monoandroid10/android-29/mcw/api.xml"
  File name: '/Users/builder/azdo/_work/3/s/xamarin-android/src/Mono.Android/obj/Release/monoandroid10/android-29/mcw/api.xml'
```

File: build-tools/api-merge/ApiDescription.cs [hunk 3]
```diff
@@ -202,9 +202,11 @@ public void Merge (string apiLocation)
 		public void Save (string filename)
 		{
 			FixupOverrides ();
-			if (filename != null)
+			if (filename != null) {
+				filename = filename.Replace ('\\', Path.DirectorySeparatorChar);
```
Comment (radekdoulik) [inline build-tools/api-merge/ApiDescription.cs]: I think this might be done earlier so that we see correct paths in the log output?

Comment (jpobst) [reply build-tools/api-merge/ApiDescription.cs]: Good call!

File: src/Mono.Android/Mono.Android.targets (hunk 2 above)
Comment (jpobst) [reply src/Mono.Android/Mono.Android.targets]: I got closer to using `$(IntermediateOutputPath)`.  The issue is that the value is for the first version it builds: `obj/Release/monoandroid10/android-29/`, and we are building for all versions, so we need to strip off the `android-29` and then generate the correct paths (`/android-28`, `/android-27`, etc.).

**Pull request merged**